### PR TITLE
rings: new package for ring graph plots

### PR DESCRIPTION
**Pull request by kortschak**
This is a wholesale import of the rings package from biogo/graphics with subsequent modifications to allow generalised use in non-genomics situations while retaining capacity for that use.

Ring plots were [first developed for use in genomics applications](http://circos.ca/) to demonstrate relationships between genomic seqments while also shown other attributes, in a concise graphic format.

For example...

There have also been many uses in non-genomics areas, for example...

I am the primary author of this code, though @brentp has made contributions. I have written to him and he is happy for me to make this move. He will need to add himself to Gonum's A+C before this is merged.

### CODE REVIEW

File: plotter/rings/axis.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// drawAt renders the axis at cen in the specified drawing area, according to the
+// Axis configuration.
+func (r *Axis) drawAt(ca draw.Canvas, cen vg.Point, fs []Scorer, base ArcOfer, inner, outer vg.Length, min, max float64) {
+	locMap := make(map[Feature]struct{})
+
+	var (
+		pa vg.Path
+
```
Comment (ctessum) [inline plotter/rings/axis.go]: Is there a reason for the double-spacing here?

File: plotter/rings/example.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
+
+package main
```
Comment (ctessum) [inline plotter/rings/example.go]: change the file name to `example_test.go` and the package name to `rings_test`.

https://blog.golang.org/examples

File: plotter/rings/example.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+			return
+		}
+	}
+	flag.Usage()
+	os.Exit(1)
+}
+
+func floatPtr(f float64) *float64 { return &f }
+
+func main() {
```
Comment (ctessum) [inline plotter/rings/example.go]: `func Example() {`

File: plotter/rings/example.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
```
Comment (ctessum) [inline plotter/rings/example.go]: delete line

File: plotter/rings/example.go (hunk 3 above)
Comment (ctessum) [reply plotter/rings/example.go]: And maybe add some documentation regarding what this is an example of?

File: plotter/rings/bezier.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
```
Comment (ctessum) [inline plotter/rings/bezier.go]: I think we already have a copy of this somewhere in the plotter package

File: plotter/rings/human.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
```
Comment (ctessum) [inline plotter/rings/human.go]: Probably make this into an official example as above.

File: plotter/rings/mouse.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
```
Comment (ctessum) [inline plotter/rings/mouse.go]: Probably make this into an official example as above

File: plotter/rings/angles.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+type ArcOfer interface {
+	Arcer
+
+	// ArcOf must return a non-nil error if the Feature is not found by
+	// the receiver or the query is nil, unless the receiver is an Arc. When
+	// the receiver is an Arc the error returned is always nil.
+	ArcOf(loc, f Feature) (Arc, error)
+}
+
+// Point represents a 2-D point.
```
Comment (ctessum) [inline plotter/rings/angles.go]: I was thinking about suggesting making this an interface instead of a type, but then I looked and I wasn't able to find anywhere in the package where it is actually used.

File: plotter/rings/example.go (hunk 2 above)
Comment (kortschak) [reply plotter/rings/example.go]: I won't do that for this file. This is residual and will go away at some point; it was used during development and helps me, but probably no-one else. There need to be a number of other examples showing each component in use, and significantly more package-level documentation.

File: plotter/rings/example.go (hunk 3 above)
Comment (kortschak) [reply plotter/rings/example.go]: See above.

File: plotter/rings/example.go (hunk 4 above)
Comment (kortschak) [reply plotter/rings/example.go]: Ditto.

File: plotter/rings/bezier.go (hunk 5 above)
Comment (kortschak) [reply plotter/rings/bezier.go]: No, this is a client of the bezier package we have. This handles the implementation of the Bézier arcs used in rings.

File: plotter/rings/human.go (hunk 6 above)
Comment (kortschak) [reply plotter/rings/human.go]: This has deps on biogo, which I don't want to bring into Gonum. I think a micro bacterial genome equivalent would work though.

File: plotter/rings/mouse.go (hunk 7 above)
Comment (kortschak) [reply plotter/rings/mouse.go]: As for human.go.

File: plotter/rings/angles.go (hunk 8 above)
Comment (kortschak) [reply plotter/rings/angles.go]: I think this is a relic. I will delete it.

File: plotter/rings/axis.go (hunk 1 above)
Comment (kortschak) [reply plotter/rings/axis.go]: I imagine there was. I'll look into what I was thinking.

Comment (kortschak) [reply plotter/rings/axis.go]: Fixed

File: plotter/rings/angles.go (hunk 8 above)
Comment (kortschak) [reply plotter/rings/angles.go]: Removed.

File: plotter/rings/labels.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			l[i] = featLabel{f}
+		}
+	}
+	return l
+}
+
+// Labels implements rendering of radial labels.
+type Labels struct {
+	// Labels contains the set of labels. Labelers that are Features and are found
+	// in the Base ArcOfer label the identified block with the string returned by
```
Comment (sbinet) [inline plotter/rings/labels.go]: this was a bit hard to parse for me: I first parsed `label` as a noun instead of as a verb.
you might want to remove this ambiguity. (but I don't feel strongly about it.)

File: plotter/rings/labels.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+	Labels []Labeler
+
+	// Base describes the ring holding the features to be labeled.
+	Base ArcOfer
+
+	// TextStyle determines the text style of each label. TextStyle behaviour
+	// is over-ridden if the Label describing a block is a TextStyler.
+	TextStyle draw.TextStyle
+
+	// Radius define the inner radius of the labels.
```
Comment (sbinet) [inline plotter/rings/labels.go]: `s/define/defines/`

File: plotter/rings/labels_test.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+			placement: func(a Angle) (rot Angle, xalign, yalign float64) {
+				return a, 0, -0.5
+			},
+		},
+		{
+			feats:     b.Set,
+			placement: Horizontal,
+		},
+	} {
+		p, err := plot.New()
```
Comment (sbinet) [inline plotter/rings/labels_test.go]: perhaps use a sub-test?

File: plotter/rings/labels_test.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+			placement: func(a Angle) (rot Angle, xalign, yalign float64) {
+				return a, 0, -0.5
+			},
+		},
+		{
+			arc:       h,
+			label:     "Label",
+			placement: Horizontal,
+		},
+	} {
```
Comment (sbinet) [inline plotter/rings/labels_test.go]: perhaps use a sub-test?

File: plotter/rings/links_test.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+			},
+		},
+		{
+			ends: [2]Feature{b.Set[0], b.Set[1]},
+			bezier: &Bezier{Segments: 5,
+				Radius: LengthDist{Length: 2 * 70 / 3, Min: floatPtr(0.95), Max: floatPtr(1.05)},
+				Crest:  &FactorDist{Factor: 2, Min: floatPtr(0.7), Max: floatPtr(1.4)},
+			},
+		},
+	} {
```
Comment (sbinet) [inline plotter/rings/links_test.go]: perhaps use a subtest?

File: plotter/rings/ribbons.go [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+	Set []Pair
+
+	// Ends holds the elements that define the end targets of the rendered ribbons.
+	Ends [2]ArcOfer
+	// Radii indicates the distance of the ribbon end points from the center of the plot.
+	Radii [2]vg.Length
+
+	// Twist indicates how feature orientation should be rendered.
+	//
+	// None indicates no explicit twist; ribbons are draw so that the start positions
```
Comment (sbinet) [inline plotter/rings/ribbons.go]: `s/draw/drawn/`

File: plotter/rings/ribbons.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+		}
+	}
+	return &Ribbons{
+		Set:   fp,
+		Ends:  ends,
+		Radii: r,
+	}, nil
+}
+
+// twist returns alters the ribbon twist depending on the relative orientation
```
Comment (sbinet) [inline plotter/rings/ribbons.go]: `s/returns alters/alters/`

File: plotter/rings/ribbons_test.go [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+							style:  blueSty,
+						},
+					},
+					sty: plotter.DefaultLineStyle,
+				},
+			},
+			segments: 5,
+			twist:    Individual | Flat,
+		},
+	} {
```
Comment (sbinet) [inline plotter/rings/ribbons_test.go]: subtest?

File: plotter/rings/rings.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+)
+
+// Twist is a flag type used to specify Ribbon and Sail twist behaviour. Specific interpretation
+// of Twist flags is documented in the relevant types.
+type Twist uint
+
+const (
+	None       Twist = 0         // None indicates no explicit twist.
+	Flat       Twist = 1 << iota // Render feature connections without twist.
+	Individual                   // Allow a feature or feature pair to define its ribbon twist.
```
Comment (sbinet) [inline plotter/rings/rings.go]: `s/Allow/Individual allows/` (for consistency?)

File: plotter/rings/rings.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+	Twisted                      // Render feature connections with twist.
```
Comment (sbinet) [inline plotter/rings/rings.go]: `s/Render/Twisted renders/` (ditto)

File: plotter/rings/rings.go [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+	End() float64
+
+	// Name returns the name of the feature.
+	Name() string
+
+	// Parent returns the reference feature on which the feature is located.
+	Parent() Feature
+}
+
+func lengthOf(f Feature) float64 {
```
Comment (sbinet) [inline plotter/rings/rings.go]: shouldn't `lengthOf` test whether `f` is `nil` and return `math.NaN`?
the doc of `Feature` says that the coordinates aren't comparable...

File: plotter/rings/rings.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+	return f.End() - f.Start()
+}
+
+// Conformationer wraps the Conformation method.
+type Conformationer interface {
+	Conformation() Conformation
+}
+
+// Conformation describes whether a feaature is linear are circular.
```
Comment (sbinet) [inline plotter/rings/rings.go]: `s/feaature/feature/`
`s/linear are/linear or/`

File: plotter/rings/rings_test.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/vg"
+	"gonum.org/v1/plot/vg/draw"
+)
+
+var regen = flag.Bool("regen", false, "Uses the current state to regenerate the test data.")
+
+func failure(f bool) string {
```
Comment (sbinet) [inline plotter/rings/rings_test.go]: remove?

File: plotter/rings/rings_test.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+		t.Errorf("image mismatch for %q", name)
+		// TODO(kortschak): Add image diffing.
+		err = ioutil.WriteFile(filepath.Join("testdata", name+"_failed.png"), got, 0664)
+		if err != nil {
+			t.Fatalf("unexpected error writing failed file: %v", err)
+		}
+	}
+}
+
+// TODO(kortschak): Use cmpimg when rings lives in plot.
```
Comment (sbinet) [inline plotter/rings/rings_test.go]: I also had to copy-paste that code in my `hplot` package...
perhaps we should just export it?

File: plotter/rings/sail.go [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+	Set []Feature
+
+	// Base holds the element that defines the end targets of the rendered sail.
+	Base ArcOfer
+	// Radius indicates the distance of the sail end points from the center of the plot.
+	Radius vg.Length
+
+	// Twist indicates how feature orientation should be rendered.
+	//
+	// None indicates no explicit twist; sail ends are draw so that the start positions
```
Comment (sbinet) [inline plotter/rings/sail.go]: `s/draw/drawn/`

File: plotter/rings/sail.go [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+	// each feature and the end positions of each feature are connected by Bézier curves.
```
Comment (sbinet) [inline plotter/rings/sail.go]: `s/each feature and/of each feature and/`

actually, I would probably rephrase as: `the start and end positions of each feature are connected ...`

File: plotter/rings/sail.go [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+	angleFeats []angleFeat
+)
+
+func (af angleFeats) Len() int { return len(af) }
+func (af angleFeats) Less(i, j int) bool {
+	return af[i].angles[0]+af[i].angles[1] < af[j].angles[0]+af[j].angles[1]
+}
+func (af angleFeats) Swap(i, j int) { af[i], af[j] = af[j], af[i] }
+
+// twist returns alters the sail twist depending on the relative orientation
```
Comment (sbinet) [inline plotter/rings/sail.go]: `s/returns alters/alters/`

File: plotter/rings/sail_test.go [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+					orient: Backward,
+					parent: b.Set[0],
+					style:  blueSty,
+				},
+			},
+			segments: 5,
+			twist:    Individual | Flat,
+		},
+	} {
+		p, err := plot.New()
```
Comment (sbinet) [inline plotter/rings/sail_test.go]: subtest?

File: plotter/rings/scale.go [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+
+// Scale represents the circular axis of ring.
+type Scale struct {
+	// Set holds a collection of features to render scales for.
+	Set []Feature
+
+	// Base defines the targets of the rendered blocks.
+	Base ArcOfer
+
+	// Radius define the radius of the axis.
```
Comment (sbinet) [inline plotter/rings/scale.go]: `s/define/defines/`

File: plotter/rings/scores.go [hunk 28]
```diff
@@ -0,0 +1,343 @@
...
+	"image/color"
+	"math"
+	"sort"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/vg"
+	"gonum.org/v1/plot/vg/draw"
+)
+
+// Scorer describes features that can provided scored values.
```
Comment (sbinet) [inline plotter/rings/scores.go]: `s/provided/provide/`

File: plotter/rings/scores.go [hunk 29]
```diff
@@ -0,0 +1,343 @@
...
+type (
+	arcScore struct {
+		Arc
+		Scorer
+	}
+	arcScores []arcScore
+)
+
+func (as arcScores) Len() int           { return len(as) }
+func (as arcScores) Less(i, j int) bool { return as[i].Theta < as[j].Theta }
```
Comment (sbinet) [inline plotter/rings/scores.go]: perhaps make sure the angles are defined in the same interval?

File: plotter/rings/scores_test.go [hunk 30]
```diff
@@ -0,0 +1,310 @@
...
+					sty := []draw.LineStyle{plotter.DefaultLineStyle, plotter.DefaultLineStyle}
+					sty[0].Color = color.NRGBA{R: 0xff, A: 0xff}
+					sty[1].Color = color.RGBA{G: 0xff, A: 0x80}
+					return sty
+				}(),
+				Join: true,
+			},
+		},
+	} {
+		p, err := plot.New()
```
Comment (sbinet) [inline plotter/rings/scores_test.go]: subtest?

File: plotter/rings/spokes.go [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+import (
+	"errors"
+	"fmt"
+
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/vg"
+	"gonum.org/v1/plot/vg/draw"
+)
+
+// Blocks implements rendering of Features representing 0 or 1 length features as radial lines.
```
Comment (sbinet) [inline plotter/rings/spokes.go]: `s/Blocks/Spokes/`

**Comment by kortschak** [comment]
As a note partly to myself, I have figured out how to integrate this as a renderer for graph types. I was originally thinking there would need to code duplication between here and graph/... to implement circular graph layout for rendering here, but I have realised that if there is an implementation of CIRCULAR (I'll add a link to the paper for this when I'm at work) in the graph packages (I'm thinking under .../graph/layout) then types satisfying a fairly minimal interface should be able to just be plotted here. I'll ponder further.

File: plotter/rings/angles.go [hunk 32]
```diff
@@ -0,0 +1,209 @@
...
+
+	return Arcs{Base: arc, Arcs: arcs}
+}
+
+// Arc returns the base arc of the Arcs.
+func (a Arcs) Arc() Arc { return a.Base }
+
+// ArcOf returns the arc of a feature in the context of the provided location.
+//
+// The behaviour of ArcOf depends on the the nil status of loc and f:
```
[the remark this answers is not in the document]
Comment (kortschak) [reply plotter/rings/angles.go]: Fixed

File: plotter/rings/axis_test.go [hunk 33]
```diff
@@ -0,0 +1,89 @@
...
+							LineStyle: plotter.DefaultLineStyle,
+							Length:    -2,
+							Label:     draw.TextStyle{Color: color.Gray16{0}, Font: font},
+						},
+					}
+				}(),
+			},
+		},
+	} {
+		p, err := plot.New()
```
[the remark this answers is not in the document]
Comment (kortschak) [reply plotter/rings/axis_test.go]: Done

File: plotter/rings/labels_test.go (hunk 11 above)
Comment (kortschak) [reply plotter/rings/labels_test.go]: Done

File: plotter/rings/labels_test.go (hunk 12 above)
Comment (kortschak) [reply plotter/rings/labels_test.go]: Done

File: plotter/rings/links_test.go (hunk 13 above)
Comment (kortschak) [reply plotter/rings/links_test.go]: Done

File: plotter/rings/ribbons.go (hunk 14 above)
Comment (kortschak) [reply plotter/rings/ribbons.go]: Fixed

File: plotter/rings/ribbons.go (hunk 15 above)
Comment (kortschak) [reply plotter/rings/ribbons.go]: Fixed

File: plotter/rings/ribbons_test.go (hunk 16 above)
Comment (kortschak) [reply plotter/rings/ribbons_test.go]: Done

File: plotter/rings/rings.go (hunk 17 above)
Comment (kortschak) [reply plotter/rings/rings.go]: Fixed

File: plotter/rings/rings.go (hunk 18 above)
Comment (kortschak) [reply plotter/rings/rings.go]: Fixed

File: plotter/rings/rings.go (hunk 19 above)
Comment (kortschak) [reply plotter/rings/rings.go]: If you have a `nil` `Feature` at this point, you have problems. The `f` here is the `Feature`, not it's parent.

File: plotter/rings/rings.go (hunk 18 above)
Comment (kortschak) [reply plotter/rings/rings.go]: I'm going to change Reverse to Invert too.

File: plotter/rings/rings.go (hunk 20 above)
Comment (kortschak) [reply plotter/rings/rings.go]: Fixed

File: plotter/rings/rings_test.go (hunk 21 above)
Comment (kortschak) [reply plotter/rings/rings_test.go]: Thanks. I was meaning to do that and forgot.

File: plotter/rings/rings_test.go (hunk 22 above)
Comment (kortschak) [reply plotter/rings/rings_test.go]: File an issue for that?

File: plotter/rings/sail.go (hunk 23 above)
Comment (kortschak) [reply plotter/rings/sail.go]: Fixed

File: plotter/rings/sail.go (hunk 24 above)
Comment (kortschak) [reply plotter/rings/sail.go]: That's nicer. Thanks.

File: plotter/rings/sail.go (hunk 25 above)
Comment (kortschak) [reply plotter/rings/sail.go]: Fixed

File: plotter/rings/sail_test.go (hunk 26 above)
Comment (kortschak) [reply plotter/rings/sail_test.go]: Done

File: plotter/rings/scale.go (hunk 27 above)
Comment (kortschak) [reply plotter/rings/scale.go]: Fixed

File: plotter/rings/scores.go (hunk 28 above)
Comment (kortschak) [reply plotter/rings/scores.go]: Fixed

File: plotter/rings/scores.go (hunk 29 above)
Comment (kortschak) [reply plotter/rings/scores.go]: That's not determinable here.

File: plotter/rings/scores_test.go (hunk 30 above)
Comment (kortschak) [reply plotter/rings/scores_test.go]: Done

File: plotter/rings/spokes.go (hunk 31 above)
Comment (kortschak) [reply plotter/rings/spokes.go]: Fixed

File: plotter/rings/rings_test.go (hunk 22 above)
Comment (sbinet) [reply plotter/rings/rings_test.go]: done: https://github.com/gonum/plot/issues/464

**Comment by kortschak** [comment]
PTAL

This needs tests on the `Lens` type and example tests for all of the types. I think it also needs more documentation in the package docs, but all these are future work.
